src/main/java/net/dries007/tfc/world/feature/vein/PipeVeinFeature.java

```diff
@@ -8,36 +8,63 @@
 
 import java.util.Random;
 
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.MutableBoundingBox;
 
 import com.mojang.serialization.Codec;
 
-public class PipeVeinFeature extends VeinFeature<PipeVeinConfig, Vein>
+public class PipeVeinFeature extends VeinFeature<PipeVeinConfig, PipeVeinFeature.PipeVein>
 {
     public PipeVeinFeature(Codec<PipeVeinConfig> codec)
     {
         super(codec);
     }
 
     @Override
-    protected MutableBoundingBox getBoundingBox(PipeVeinConfig config)
+    protected MutableBoundingBox getBoundingBox(PipeVeinConfig config, PipeVein vein)
     {
-        return new MutableBoundingBox(-config.getRadius(), -config.getSize(), -config.getRadius(), config.getRadius(), config.getSize(), config.getRadius());
+        int radius = config.getRadius();
+        int skew = vein.skew;
+        return new MutableBoundingBox(-radius - skew, -config.getSize(), -radius - skew, radius + skew, config.getSize(), radius + skew);
     }
 
     @Override
-    protected float getChanceToGenerate(int x, int y, int z, Vein vein, PipeVeinConfig config)
+    protected float getChanceToGenerate(int x, int y, int z, PipeVein vein, PipeVeinConfig config)
     {
-        if (Math.abs(y) < config.getSize() && (x * x) + (z * z) < config.getRadius() * config.getRadius())
+        final double yScaled = (double) y / config.getSize();
+        x += vein.skew * MathHelper.cos(vein.angle) * yScaled;
+        z += vein.skew * MathHelper.sin(vein.angle) * yScaled;
+
+        final double yFactor = (double) vein.yFlipper * yScaled + 0.5D;
+        final double trueRadius = config.getRadius() * (1 - yFactor) + (config.getRadius() - vein.slant) * yFactor;
+        if (Math.abs(y) < config.getSize() && (x * x) + (z * z) < trueRadius * trueRadius)
         {
             return config.getDensity();
         }
         return 0;
     }
 
     @Override
-    protected Vein createVein(int chunkX, int chunkZ, Random random, PipeVeinConfig config)
+    protected PipeVein createVein(int chunkX, int chunkZ, Random random, PipeVeinConfig config)
+    {
+        return new PipeVein(defaultPos(chunkX, chunkZ, random, config), random, config);
+    }
+
+    static class PipeVein extends Vein
     {
-        return new Vein(defaultPos(chunkX, chunkZ, random, config));
+        final int yFlipper;
+        final float angle;
+        final int skew;
+        final int slant;
+
+        PipeVein(BlockPos pos, Random random, PipeVeinConfig config)
+        {
+            super(pos);
+            this.yFlipper = random.nextBoolean() ? 1 : -1;
```

Comment:
This should
a. be called `sign` not `yFlipper` :P
b. be a field in json - `sign`, which is a float between 0 and 1. Then `vein.sign = type.sign < random.nextFloat() ? -1 : 1`

- If `sign = 0`, then it will always set to 1
- If `sign = 1`, then it will always set to -1
- Anywhere between is a linear probability between the two.

Preferably document so we know what the probability is actually of - i.e. which cone shape.